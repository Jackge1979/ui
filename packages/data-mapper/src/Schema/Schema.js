import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import TableRenderer from './TableRenderer';
import Constants from '../Constants';

function isMapped(dataAccessor, element, mappedElements) {
	return mappedElements == null ? false : dataAccessor.includes(mappedElements, element);
}

function isHighlighted(dataAccessor, element, selection, side, pendingItem, focusedElements, dnd) {
	const connected =
		selection == null
			? false
			: selection.side !== side &&
			  selection.connected != null &&
			  dataAccessor.includes(selection.connected, element);
	const pending =
		pendingItem != null &&
		pendingItem.side === side &&
		dataAccessor.areElementsEqual(pendingItem.element, element);
	const focused = focusedElements != null && dataAccessor.includes(focusedElements, element);
	const isTarget = dnd && dnd.target && dataAccessor.areElementsEqual(dnd.target.element, element);
	return connected || pending || focused || isTarget;
}

export default class Schema extends Component {
	constructor(props) {
		super(props);
		this.updateRendererNodeRef = this.updateRendererNodeRef.bind(this);
		this.onContentScroll = this.onContentScroll.bind(this);
	}

	shouldComponentUpdate(nextProps) {
		// check first is a drag and drop is in progress
		let needUpdate = true;
		if (nextProps.dnd) {
			needUpdate = !(nextProps.dnd.source != null && nextProps.dnd.source.side === nextProps.side);
		}
		// then check if rendering focused elements is needed
		if (
			needUpdate &&
			nextProps.trigger &&
			(nextProps.trigger.code === Constants.Events.ENTER_ELEM ||
				nextProps.trigger.code === Constants.Events.LEAVE_ELEM)
		) {
			if (
				this.props.dataAccessor.haveSameContent(
					this.props.focusedElements,
					nextProps.focusedElements,
				)
			) {
				needUpdate = false;
			}
		}
		return needUpdate;
	}

	getNode(element) {
		const index = this.props.dataAccessor.getSchemaElementIndex(this.props.schema, element, true);
		const children = this.getRendererNode().getChildNodes();
		const childrenArray = Array.from(children);
		return childrenArray[index];
	}

	getYPosition(element) {
		const scrollTop = this.getRendererNode().getScrollTop();
		const child = this.getNode(element);
		if (child) {
			const childOffsetTop = this.getRendererNode().getChildOffsetTop(child);
			return childOffsetTop + child.clientHeight / 2 - scrollTop;
		}
		return 0;
	}

	getElementAtPosition(position) {
		let theElement = null;
		if (this.isRendererNodeDefined()) {
			let previousDist = -1;
			let currentDist = -1;
			const contentHeight = this.getRendererNode().getOffsetHeight();
			const elements = this.props.dataAccessor.getSchemaElements(this.props.schema, true);
			const children = this.getRendererNode().getChildNodes();
			const childrenArray = Array.from(children);
			for (let i = 0; i < childrenArray.length; i += 1) {
				previousDist = currentDist;
				const element = elements[i];
				const elemYPos = this.getYPosition(element);
				if (elemYPos > 0 && elemYPos < contentHeight) {
					currentDist = Math.abs(elemYPos - position);
					if (previousDist >= 0 && currentDist > previousDist) {
						break;
					}
					theElement = element;
				} else if (elemYPos > contentHeight) {
					break;
				}
			}
		}
		return theElement;
	}

	onContentScroll() {
		this.props.onScroll(this.props.side);
	}

	getVisibleElements() {
		let visibleElements = [];
		if (this.isRendererNodeDefined()) {
			const contentHeight = this.getRendererNode().getOffsetHeight();
			const elements = this.props.dataAccessor.getSchemaElements(this.props.schema, true);
			const children = this.getRendererNode().getChildNodes();
			const childrenArray = Array.from(children);
			if (elements.length != childrenArray.length) {
				visibleElements = visibleElements.concat(elements);
				return visibleElements;
			}
			const meanDist = this.computeMeanDist(elements);
			let startIndex = 0;
			if (meanDist > 0) {
				// compute start index
				const scrollTop = this.getRendererNode().getScrollTop();
				const n = Math.floor(scrollTop / meanDist);
				startIndex = Math.max(0, n);
			}
			const headerHeight = this.getRendererNode().getHeaderHeight();
			const bottomLimit = contentHeight + headerHeight;
			for (let i = startIndex; i < childrenArray.length; i += 1) {
				const element = elements[i];
				const elemYPos = this.getYPosition(element);
				if (elemYPos > headerHeight && elemYPos < bottomLimit) {
					// element is visible
					visibleElements = visibleElements.concat(element);
				} else if (elemYPos > bottomLimit) {
					break;
				}
			}
		}
		return visibleElements;
	}

	computeMeanDist(elements) {
		if (elements.length > 1) {
			const y1 = this.getYPosition(elements[0]);
			const y2 = this.getYPosition(elements[1]);
			return Math.abs(y2 - y1);
		}
		return 0;
	}

	reveal(element) {
		const node = this.getNode(element);
		const nodeHeight = node.clientHeight;
		const elemYPos = this.getYPosition(element);
		const contentHeight = this.getRendererNode().getOffsetHeight();
		const headerHeight = this.getRendererNode().getHeaderHeight();
		let revealed = false;
		const scrollTop = this.getRendererNode().getScrollTop();
		if (elemYPos < headerHeight) {
			const newScrollTop = scrollTop + elemYPos - headerHeight - nodeHeight / 2;
			this.getRendererNode().setScrollTop(newScrollTop);
			revealed = true;
		} else if (elemYPos > contentHeight + headerHeight - nodeHeight) {
			const offset = elemYPos + nodeHeight - contentHeight - headerHeight;
			this.getRendererNode().setScrollTop(scrollTop + offset);
			revealed = true;
		}
		return revealed;
	}

	isRendererNodeDefined() {
		return this.rendererNode;
	}

	getRendererNode() {
		if (this.rendererNode) {
			return this.rendererNode;
		}
		return {
			getChildNodes() {
				return [];
			},
			getScrollTop() {
				return 0;
			},
			setScrollTop(scrollTop) {},
			getChildOffsetTop(child) {
				return 0;
			},
			getOffsetHeight() {
				return 0;
			},
			getHeaderHeight() {
				return 0;
			},
		};
	}

	updateRendererNodeRef(ref) {
		this.rendererNode = ref;
	}

	render() {
		const { withTitle, ...tempProps } = this.props;
		const { dataAccessor, schema, side } = this.props;
		const contentProps = {
			...tempProps,
			isMapped,
			isHighlighted,
			onScroll: this.onContentScroll,
		};		
		return (
			<div className={`schema mapper-element ${side}`}>
				<TableRenderer
					ref={this.updateRendererNodeRef}
					title={withTitle && dataAccessor.getSchemaName(schema)}
					{...contentProps}
				/>
			</div>
		);
	}
}

Schema.propTypes = {
	dataAccessor: PropTypes.object.isRequired,
	schema: PropTypes.object.isRequired,
	columns: PropTypes.array.isRequired,
	rowsClassName: PropTypes.objectOf(PropTypes.string),
	withTitle: PropTypes.bool,
	withHeader: PropTypes.bool,
	filters: PropTypes.array,
	onFilterChange: PropTypes.func,
	sorters: PropTypes.object,
	onSortChange: PropTypes.func,
	focusedElements: PropTypes.array,
	onScroll: PropTypes.func,
	side: PropTypes.string,
	isElementVisible: PropTypes.func,
};
