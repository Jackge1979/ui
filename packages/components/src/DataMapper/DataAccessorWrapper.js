import * as Constants from './Constants';

export default class DataAccessorWrapper {

  constructor(dataAccessor) {
    this.dataAccessor = dataAccessor;
  }

  /**
	* Returns the name of the schema.
	*/
	getSchemaName(schema) {
		return this.dataAccessor.getSchemaName(schema);
	}

	/**
	* Returns the number of elements in the schema.
	*/
	getSchemaSize(schema) {
		return this.dataAccessor.getSchemaSize(schema);
	}

	/**
	* Returns all the elements of the schema in an array.
	*/
	getSchemaElements(schema) {
		return this.dataAccessor.getSchemaElements(schema);
	}

	/**
	* Returns the nth element of the schema.
	*/
	getSchemaElement(schema, index) {
    if (this.dataAccessor.getSchemaElement) {
      return this.dataAccessor.getSchemaElement(schema, index);
    }
		const elements = this.getSchemaElements(schema);
    if (elements && elements.length > index) {
      return elements[index];
    }
    return null;
	}

	/**
	* Returns the index of the given element,
	* returns -1 if it is not in the schema.
	*/
	getSchemaElementIndex(schema, element) {
    if (this.dataAccessor.getSchemaElementIndex) {
      return this.dataAccessor.getSchemaElementIndex(schema, element);
    }
    const elements = this.getSchemaElements(schema);
    if (elements) {
      return elements.indexOf(element);
    }
		return -1;
	}

  getSchemaElementFromId(schema, id) {
    if (this.dataAccessor.getSchemaElementFromId) {
      return this.dataAccessor.getSchemaElementFromId(schema, id);
    }
    const elements = this.getSchemaElements(schema);
    if (elements) {
      return elements.find(elem => this.getElementId(elem) === id);
    }
    return null;
  }

	/**
	* Returns the identifier of the element.
	* Identifier must be unique.
	*/
	getElementId(element) {
		return this.dataAccessor.getElementId(element);
	}

	/**
	* Returns true if the two elements are equals.
	* Default implementation is based on element id.
	*/
	areEquals(element1, element2) {
    if (this.dataAccessor.areEquals) {
      return this.dataAccessor.areEquals(element1, element2);
    }
		return this.getElementId(element1) === this.getElementId(element2);
	}

	/**
	* Returns true if the array of elements contains the specified element.
	*/
	includes(elements, element) {
    if (this.dataAccessor.includes) {
      return this.dataAccessor.includes(elements, element);
    }
    for (let i = 0; i < elements.length; i += 1) {
      if (this.areEquals(elements[i], element)) {
        return true;
      }
    }
    return false;
	}

	/**
	* Returns the name of the element.
	*/
	getElementName(element) {
		return this.dataAccessor.getElementName(element);
	}

  haveSameName(element1, element2) {
  	return this.getElementName(element1) === this.getElementName(element2);
  }

	/**
	* Returns the semantic type of the element.
	*/
	getElementType(element) {
		return this.dataAccessor.getElementType(element);
	}

	/**
	* Returns a description fot the element.
	*/
	getElementDescription(element) {
		return this.dataAccessor.getElementDescription(element);
	}

	/**
	* Returns an array of all the mapping items. A mapping item defines
	* a mapping betwwen an input and an output element.
	*/
	getMappingItems(mapping) {
		return this.dataAccessor.getMappingItems(mapping);
	}

  /**
  * Returns true if the given mapping is empty
  */
  isMappingEmpty(mapping) {
    if (this.dataAccessor.isMappingEmpty) {
      return this.dataAccessor.isMappingEmpty(mapping);
    }
    const items = this.getMappingItems(mapping);
    return !items || !items.length;
  }

	/**
	* Return the mapped element for the specified side.
	*/
	getMappedElement(mappingItem, side) {
		return this.dataAccessor.getMappedElement(mappingItem, side);
	}

	/**
	* Add a new mapping from source element to target element.
	* Returns the updated mapping.
	*/
	addMapping(mapping, source, target) {
		return this.dataAccessor.addMapping(mapping, source, target);
	}

	/**
	* Remove the (source->target) mapping.
	* Returns the updated mapping.
	*/
	removeMapping(mapping, source, target) {
		return this.dataAccessor.removeMapping(mapping, source, target);
	}

	/**
	* Remove all the mapping items.
	* Returns the updated mapping.
	*/
	clearMapping(mapping) {
		return this.dataAccessor.clearMapping(mapping);
	}

  // Some more convenient methods

  isElementInMappingItem(mappingItem, element, side) {
  	const mappedElement = this.getMappedElement(mappingItem, side);
  	return this.areEquals(mappedElement, element);
  }

  /**
   * isElementMapped returns true if the given (element, side) is mapped
   * (i.e. if it appears in the mapping)
   */
  isElementMapped(mapping, element, side) {
  	if (mapping != null) {
  		const mappingItems = this.getMappingItems(mapping);
  		return mappingItems.find(
  			item => this.isElementInMappingItem(item, element, side)
  		);
  	}
  	return false;
  }

  /**
  * isFullMapped returns true if all the elements of the given schema are mapped
  */
  isFullMapped(mapping, schema, side) {
  	// TODO could be optimized
  	for (let i = 0; i < this.getSchemaSize(schema); i += 1) {
  		if (!this.isElementMapped(mapping, this.getSchemaElement(schema, i), side)) {
  			return false;
  		}
  	}
  	return true;
  }

  getMappingItemsWithElement(mapping, element, side) {
  	const mappingItems = this.getMappingItems(mapping);
  	return mappingItems.filter(item => this.isElementInMappingItem(item, element, side));
  }

  getConnectedElements(mapping, element, side) {
  	const items = this.getMappingItemsWithElement(mapping, element, side);
  	if (items != null) {
  		const connectedSide = Constants.switchMappingSide(side);
  		return items.map(item => this.getMappedElement(item, connectedSide));
  	}
  	return null;
  }

}
