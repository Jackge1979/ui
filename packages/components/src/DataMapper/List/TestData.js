import RowLabel from './RowLabel';
import MandatoryField from './MandatoryField';

export const element1 = {
	id: 'elem_1',
	name: 'element 1',
	type: 'string',
	description: 'bla bla bla',
  mandatory: false,
};

export const element2 = {
	id: 'elem_2',
	name: 'element 2',
	type: 'string',
	description: 'This element is mandatory',
  mandatory: true,
};

export const KEYS = {
  NAME: 'name',
  TYPE: 'type',
  DESC: 'desc',
};

export const dataKeys = [KEYS.NAME, KEYS.TYPE, KEYS.DESC];

export const classNameProvider = {
  get(element, key) {
    if (element && key) {
			return `classname-of-(${element.id},${key})`;
		} else if (element) {
			return `classname-of-(${element.id})`;
		}
		return `classname-of-list`;
  },
  onDragOver(status) {
  },
};

export const rowDataGetter = {
  getId(element) {
    return element.id;
  },
  getData(element, key) {
		switch (key) {
			case KEYS.NAME:
				return {
					value: element.name,
					mandatory: element.mandatory,
				};
			case KEYS.TYPE:
				return element.type;
			case KEYS.DESC:
				return element.description;
			default:
				return `No data available for ${key}`;
		}
	},
};

export const rowRenderers = {
  getComponent(key) {
		switch (key) {
			case KEYS.NAME:
				return MandatoryField;
			default:
				return RowLabel;
		}
	},
};
