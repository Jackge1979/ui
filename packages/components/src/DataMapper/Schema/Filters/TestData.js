export const Filters = {
	WITH: true,
	WITHOUT: false,
};

export const element1 = {
	id: 'elem_1',
	name: 'Firstname',
	type: 'string',
	description: 'bla bla bla',
	mandatory: false,
};

export const element2 = {
	id: 'elem_2',
	name: 'Lastname',
	type: 'string',
	description: 'bla bla bla',
	mandatory: true,
};

export const element3 = {
	id: 'elem_3',
	name: 'Birthday',
	type: 'date',
	description: 'bla bla bla',
	mandatory: false,
};

export const element4 = {
	id: 'elem_4',
	name: 'Address',
	type: 'address',
	description: 'bla bla bla',
	mandatory: true,
};

export const schema = {
	id: '5df4g6ds4df564s',
	name: 'schema',
	elements: [element1, element2, element3, element4],
};
