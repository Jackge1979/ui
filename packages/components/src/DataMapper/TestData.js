
export const element1 = {
	id: 'elem_1',
	name: 'Firstname',
	type: 'string',
	description: 'The firstname of the customer (optional)',
	mandatory: false,
};

export const element2 = {
	id: 'elem_2',
	name: 'Lastname',
	type: 'string',
	description: 'The lastname of the customer (mandatory)',
	mandatory: true,
};

export const element3 = {
	id: 'elem_3',
	name: 'Birthday',
	type: 'date',
	description: 'Thje birthday of the customer (optional)',
	mandatory: false,
};

export const element4 = {
	id: 'elem_4',
	name: 'Address',
	type: 'address',
	description: 'The address of the customer (mandatory)',
	mandatory: true,
};

export const schema = {
	id: '5df4g6ds4df564s',
	name: 'Customer info',
	elements: [element1, element2, element3, element4],
};
