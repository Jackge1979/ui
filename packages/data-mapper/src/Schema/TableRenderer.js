import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Table, DraggableComponent as draggable } from '@talend/react-components';

function getRowsClassNames(rowsClassName, side, elements, dataAccessor, schemaProps) {
	const {
		selection,
		pendingItem,
		dnd,
		focusedElements,
		mappedElements,
		isHighlighted,
		isMapped,
		isSelected,
	} = schemaProps;
	const rowsClassNames = {};
	for (let i = 0; i < elements.length; i += 1) {
		const element = elements[i];
		const elementId = dataAccessor.getElementId(element);
		rowsClassNames[elementId] = classnames(
			{
				highlighted: isHighlighted(
					dataAccessor,
					element,
					selection,
					side,
					pendingItem,
					focusedElements,
					dnd,
				),
				mapped: isMapped(dataAccessor, element, mappedElements),
				selected: isSelected(dataAccessor, selection, element, side),
			},
			rowsClassName && rowsClassName[elementId],
		);
	}
	return rowsClassNames;
}

// function updateClassNames(classNames, side, elements, dataAccessor, schemaProps) {
// 	return {
// 		root: classNames && classNames.root,
// 		titleBar: classNames && classNames.titleBar,
// 		title: classNames && classNames.title,
// 		filtersBar: classNames && classNames.filtersBar,
// 		table: classnames('schema-content', classNames && classNames.table),
// 		header: classNames && classNames.header,
// 		body: classNames && classNames.body,
// 		row: classnames(classNames && classNames.row, 'draggable-row'),
// 		rows: getRowsClassNames(classNames, side, elements, dataAccessor, schemaProps),
// 	};
// }

function copyColumn(column) {
	const newColumn = {};
	for (let k in column) {
		newColumn[k] = column[k];
	}
	return newColumn;
}

function copyColumns(columns) {
	let result = [];
	for (let i = 0; i < columns.length; i += 1) {
		result = result.concat(copyColumn(columns[i]));
	}
	return result;
}

class InternalDndListener {
	constructor() {
		this.beginDrag = this.beginDrag.bind(this);
		this.canDrop = this.canDrop.bind(this);
		this.drop = this.drop.bind(this);
		this.endDrag = this.endDrag.bind(this);
	}

	update(schemaProps) {
		this.schemaProps = schemaProps;
	}

	beginDrag(element) {
		return this.schemaProps.dndListener.beginDrag(element, this.schemaProps.side);
	}

	canDrop(sourceItem, targetElement) {
		const targetItem = { element: targetElement, side: this.schemaProps.side };
		return this.schemaProps.dndListener.canDrop(sourceItem, targetItem);
	}

	drop(sourceItem, targetElement) {
		const targetItem = { element: targetElement, side: this.schemaProps.side };
		this.schemaProps.dndListener.drop(sourceItem, targetItem);
	}

	endDrag() {
		this.schemaProps.dndListener.endDrag();
	}
}

class InternalSelectionHandler {
	constructor() {
		this.onClick = this.onClick.bind(this);
		this.onDoubleClick = this.onDoubleClick.bind(this);
	}

	update(schemaProps) {
		this.schemaProps = schemaProps;
	}

	onClick(element, ev) {
		this.schemaProps.onSelect(ev.ctrlKey, element, this.schemaProps.side);
	}

	onDoubleClick(element) {
		this.schemaProps.revealConnectedElement(element, this.schemaProps.side);
	}
}

class ColumnUpdater {
	constructor() {
		this.dndListener = new InternalDndListener();
		this.selectionHandler = new InternalSelectionHandler();
		this.draggableCell = null;
	}

	update(schemaProps) {
		this.schemaProps = schemaProps;
		this.dndListener.update(schemaProps);
		this.selectionHandler.update(schemaProps);
	}

	// updateClassNameWithIO(className) {
	// 	// specific input/output className added for data-mapper context
	// 	const classes = {
	// 		input: this.schemaProps.side === Constants.MappingSide.INPUT,
	// 		output: this.schemaProps.side === Constants.MappingSide.OUTPUT,
	// 	};
	// 	return classnames(classes, className);
	// }

	addDnd(column) {
		if (!this.draggableCell) {
			this.draggableCell = draggable(column.cellRenderer || Table.Cell, 'element');
		}
		column.cellRenderer = this.draggableCell;
		// add dnd callback methods
		column.cellExtraProps = {};
		column.cellExtraProps.beginDrag = this.dndListener.beginDrag;
		column.cellExtraProps.canDrop = this.dndListener.canDrop;
		column.cellExtraProps.drop = this.dndListener.drop;
		column.cellExtraProps.endDrag = this.dndListener.endDrag;
	}

	addSelection(column) {
		column.cellExtraProps.onClick = this.selectionHandler.onClick;
		column.cellExtraProps.onDoubleClick = this.selectionHandler.onDoubleClick;
	}

	updateColumns(columns) {
		const columnsWithDnd = copyColumns(columns);
		// update columns classnames
		// for (let i = 0; i < columnsWithDnd.length; i += 1) {
		// 	columnsWithDnd[i].headClassName = this.updateClassNameWithIO(columnsWithDnd[i].headClassName);
		// 	columnsWithDnd[i].cellClassName = this.updateClassNameWithIO(columnsWithDnd[i].cellClassName);
		// }
		// add dnd baheviour on the first column
		this.addDnd(columnsWithDnd[0]);
		// add selection behaviour on the first column
		this.addSelection(columnsWithDnd[0]);
		return columnsWithDnd;
	}

}

class TableRenderingListener {

	onMounted(part, node) {
		this.updateRef(part, node);
	}

	onUpdated(part, node) {
		this.updateRef(part, node);
	}

	updateRef(part, node) {
		switch (part) {
			case 'table':
				this.tableNode = node;
				break;
			case 'head':
				this.headNode = node;
				break;
			case 'body':
				this.bodyNode = node;
				break;
			default:
				break;
		}
	}

	getTableNode() {
		return this.tableNode;
	}

	getHeadNode() {
		return this.headNode;
	}

	getBodyNode() {
		return this.bodyNode;
	}
}

export default class TableRenderer extends Component {
	constructor(props) {
		super(props);
		this.columnUpdater = new ColumnUpdater();
		this.renderingListener = new TableRenderingListener();
		this.onEnterElement = this.onEnterElement.bind(this);
		this.onLeaveElement = this.onLeaveElement.bind(this);
		this.onFilterChange = this.onFilterChange.bind(this);
		this.onSortChange = this.onSortChange.bind(this);
	}

	onEnterElement(element) {
		this.props.onEnterElement(element, this.props.side);
	}

	onLeaveElement(element) {
		this.props.onLeaveElement(element, this.props.side);
	}

	onFilterChange(id, active, params) {
		this.props.onFilterChange(id, active, params, this.props.side);
	}

	onSortChange(columnKey) {
		this.props.onSortChange(columnKey, this.props.side);
	}

	getChildNodes() {
		return this.renderingListener.getBodyNode().childNodes;
	}

	getScrollTop() {
		return this.renderingListener.getBodyNode().scrollTop;
	}

	setScrollTop(scrollTop) {
		this.renderingListener.getBodyNode().scrollTop = scrollTop;
	}

	getChildOffsetTop(child) {
		const childOffsetTop = child.offsetTop;
		const bodyOffsetTop = this.renderingListener.getBodyNode().offsetTop;
		const tableOffsetTop = this.renderingListener.getTableNode().offsetTop;
		return childOffsetTop;
	}

	getOffsetHeight() {
		return this.renderingListener.getBodyNode().offsetHeight;
	}

	getHeaderHeight() {
		return this.renderingListener.getHeadNode().offsetHeight;
	}

	render() {
		this.columnUpdater.update(this.props);
		const {
			dataAccessor,
			schema,
			onScroll,
			columns,
			rowsClassName,
			withHeader,
			filters,
			sorters,
			title,
			side,
		} = this.props;
		const elements = dataAccessor.getSchemaElements(schema, true);
		//const tableClassNames = updateClassNames(classNames, side, elements, dataAccessor, this.props)
		const columnsWithDnd = this.columnUpdater.updateColumns(columns);
		return (
			<Table
				title={title}
				elements={elements}
				columns={columnsWithDnd}
				rowsClassName={getRowsClassNames(rowsClassName, side, elements, dataAccessor, this.props)}
				withHeader={withHeader}
				filters={filters}
				onFilterChange={this.onFilterChange}
				sorters={sorters}
				onSortChange={this.onSortChange}
				onScroll={onScroll}
				onEnterRow={this.onEnterElement}
				onLeaveRow={this.onLeaveElement}
				renderingListener={this.renderingListener}
			/>
		);
	}
}

TableRenderer.propTypes = {
	dataAccessor: PropTypes.object,
	schema: PropTypes.object,
	title: PropTypes.string,
	columns: PropTypes.array,
	rowsClassName: PropTypes.objectOf(PropTypes.string),
	withHeader: PropTypes.bool,
	filters: PropTypes.array,
	onFilterChange: PropTypes.func,
	sorters: PropTypes.object,
	onSortChange: PropTypes.func,
	onScroll: PropTypes.func,
	side: PropTypes.string,
	onSelect: PropTypes.func,
	revealConnectedElement: PropTypes.func,
	onEnterElement: PropTypes.func,
	onLeaveElement: PropTypes.func,
	isElementVisible: PropTypes.func,
};
