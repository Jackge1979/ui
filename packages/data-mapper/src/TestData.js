export const emptySchema = {
	id: 'empty_schema',
	name: 'Empty schema',
	elements: [],
};

export const element1 = {
	id: 'elem_1',
	name: 'Firstname',
	type: 'string',
	description: 'The firstname of the customer (optional)',
	mandatory: false,
};

export const element2 = {
	id: 'elem_2',
	name: 'Lastname',
	type: 'string',
	description: 'The lastname of the customer (mandatory)',
	mandatory: true,
};

export const element3 = {
	id: 'elem_3',
	name: 'Birthday',
	type: 'date',
	description: 'The birthday of the customer (optional)',
	mandatory: false,
};

export const element4 = {
	id: 'elem_4',
	name: 'Address',
	type: 'address',
	description: 'The address of the customer (mandatory)',
	mandatory: true,
};

export const schema1 = {
	id: '5df4g6ds4df564s',
	name: 'Customer info',
	elements: [element1, element2, element3, element4],
};

export const element5 = {
	id: 'elem_5',
	name: 'Name',
	type: 'string',
	description: 'The name of the client',
	mandatory: true,
};

export const element6 = {
	id: 'elem_6',
	name: 'Role',
	type: 'string',
	description: 'The role of the client',
	mandatory: false,
};

export const element7 = {
	id: 'elem_7',
	name: 'Birthday',
	type: 'date',
	description: 'The birthday of the client',
	mandatory: false,
};

export const element8 = {
	id: 'elem_8',
	name: 'Address',
	type: 'address',
	description: 'The address of the client',
	mandatory: true,
};

export const schema2 = {
	id: '1bf1dbw1d81w51',
	name: 'Client data',
	elements: [element5, element6, element7, element8],
};

export const mapping = [
	{
		source: element2,
		target: element5,
	},
	{
		source: element4,
		target: element8,
	},
];

export const Columns = {
	NAME: {
		id: 'name-col-id',
		key: 'name',
		label: 'Name',
	},
	TYPE: {
		id: 'type-col-id',
		key: 'type',
		label: 'Type',
	},
	DESC: {
		id: 'desc-col-id',
		key: 'description',
		label: 'Description',
	},
	MANDATORY: {
		id: 'mand-col-id',
		key: 'mandatory',
		label: '',
	},
};
