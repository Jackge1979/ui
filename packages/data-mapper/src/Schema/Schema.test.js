import React from 'react';
import renderer from 'react-test-renderer';
import Schema from './Schema.js';
import DataAccessorWrapper from '../DataAccessor/DataAccessorWrapper';
import * as TestData from '../TestData';
import { Constants } from '../index';

const columns = [TestData.Columns.NAME, TestData.Columns.TYPE, TestData.Columns.DESC];

describe('Schema', () => {
	it('should accept a single schema', () => {
		const dataAccessor = new DataAccessorWrapper();
		dataAccessor.registerSchema(TestData.schema1, Constants.MappingSide.INPUT);
		// create React tree
		const tree = renderer
			.create(<Schema dataAccessor={dataAccessor} schema={TestData.schema1} columns={columns} />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('empty-schema', () => {
		const dataAccessor = new DataAccessorWrapper();
		dataAccessor.registerSchema(TestData.emptySchema, Constants.MappingSide.INPUT);
		// create React tree
		const tree = renderer
			.create(
				<Schema dataAccessor={dataAccessor} schema={TestData.emptySchema} columns={columns} />,
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
