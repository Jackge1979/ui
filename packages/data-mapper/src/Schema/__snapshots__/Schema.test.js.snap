// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema empty-schema 1`] = `
<div
  className="schema mapper-element undefined"
>
  <div
    className="tc-table theme-tc-table"
  >
    <table>
      <tbody
        className="tc-table-body theme-tc-table-body"
        onScroll={[Function]}
      />
    </table>
  </div>
</div>
`;

exports[`Schema should accept a single schema 1`] = `
<div
  className="schema mapper-element undefined"
>
  <div
    className="tc-table theme-tc-table"
  >
    <table>
      <tbody
        className="tc-table-body theme-tc-table-body"
        onScroll={[Function]}
      >
        <tr
          className="tc-table-row theme-tc-table-row"
          data-id="elem_1"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <td
            className="td-name theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="draggable-component"
            >
              <div
                className="tc-table-clickable-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-name"
                onClick={[Function]}
                onDoubleClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="-1"
              >
                Firstname
              </div>
            </div>
          </td>
          <td
            className="td-type theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="tc-table-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-type"
            >
              string
            </div>
          </td>
          <td
            className="td-description theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="tc-table-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-description"
            >
              The firstname of the customer (optional)
            </div>
          </td>
        </tr>
        <tr
          className="tc-table-row theme-tc-table-row"
          data-id="elem_2"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <td
            className="td-name theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="draggable-component"
            >
              <div
                className="tc-table-clickable-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-name"
                onClick={[Function]}
                onDoubleClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="-1"
              >
                Lastname
              </div>
            </div>
          </td>
          <td
            className="td-type theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="tc-table-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-type"
            >
              string
            </div>
          </td>
          <td
            className="td-description theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="tc-table-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-description"
            >
              The lastname of the customer (mandatory)
            </div>
          </td>
        </tr>
        <tr
          className="tc-table-row theme-tc-table-row"
          data-id="elem_3"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <td
            className="td-name theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="draggable-component"
            >
              <div
                className="tc-table-clickable-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-name"
                onClick={[Function]}
                onDoubleClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="-1"
              >
                Birthday
              </div>
            </div>
          </td>
          <td
            className="td-type theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="tc-table-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-type"
            >
              date
            </div>
          </td>
          <td
            className="td-description theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="tc-table-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-description"
            >
              The birthday of the customer (optional)
            </div>
          </td>
        </tr>
        <tr
          className="tc-table-row theme-tc-table-row"
          data-id="elem_4"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <td
            className="td-name theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="draggable-component"
            >
              <div
                className="tc-table-clickable-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-name"
                onClick={[Function]}
                onDoubleClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="-1"
              >
                Address
              </div>
            </div>
          </td>
          <td
            className="td-type theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="tc-table-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-type"
            >
              address
            </div>
          </td>
          <td
            className="td-description theme-tc-table-row-cell"
            headers={undefined}
          >
            <div
              className="tc-table-cell tc-table-row-data theme-tc-table-row-data tc-table-row-data-description"
            >
              The address of the customer (mandatory)
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
